### ISSUE TITLE
Bibtool as a linter

### ISSUE THREAD

**Issue by kiryph** [ge-ne/bibtool#52]
Many (modern) text editors support lint like tools. They usually collect the returned messages in a separate window (in vim it is called the quickfix window). Futhermore line markers are added to automatically inform the user without too much distraction about possible issues.

In the vim plugin world there exist two popular ones for this purpose:
1. https://github.com/w0rp/ale
2. https://github.com/vim-syntastic/syntastic

`Ale` right now uses `bibclean` for bib files.

I was wondering whether bibtool could also be used or modified that it can be used for this?

**Comment by ge-ne** [comment]
BibTool contains a parer which emits error messages as required. You can redirect the otput to /dev/null (on a UNIXoid OS). The exit code signals success or error. I could imagine to add an option to suppress output.
What is required to use BibTool as linter?

**Comment by kiryph** [comment]
I think redirecting to `/dev/null` is sufficient. However, I am not sure about MS Windows.

My request is more or less about line numbers:

[`bibclean`](http://ftp.math.utah.edu/pub/bibclean/) ([ctan](https://ctan.org/pkg/bibclean)) can already be used as a linter. The output looks like

```
$ bibclean my.bib > /dev/null
%% my.bib:6:Expected http://dx.doi.org/ prefix in DOI value ``"10.1098/rstl.1856.0022"''.
%% my.bib:10:Unexpected value in ``month = "1"''.
%% my.bib:18:Expected http://dx.doi.org/ prefix in DOI value ``"10.1080/14786448708628135"''.
%% my.bib:23:Unexpected protocol://... in URL value ``"https://doi.org/10.1080/14786448708628135"'': normally ftp://... or http://....
%% my.bib:23:Possible DOI in URL value ``"https://doi.org/10.1080/14786448708628135"'': if so, move to separate DOI = "..." key/value in this entry.
```

The format of each message is `%% filename:line number:message`. So BibTool should print similar messages.
`bibclean` can actually also print column numbers which would allow to place the cursor at the exact position. However, from my point of view line numbers are good enough. If the BibTool code already contains a column number it would be nice to print it as well, but not a deal breaker.

To make this more explicit, consider this short bib file:

```
@article{Koch:1984,
  author  = {Koch, E.},
  doi     = {10.1107/S0108767384001227},
  issn    = {0108-7673},
  journal = {Acta Crystallographica Section A Foundations of Crystallography},
  month   = {sep},
  number  = {5},
  pages   = {593--600},
  title   = {The implications of normalizers on group–subgroup relations between space groups},
  volume  = {40},
  year    = {1984}
}
@Article{Dove:1997,
  author  = {Dove, Martin T.},
  title   = {Theory of displacive phase transitions in minerals},
  doi     = {10.2138/am-1997-3-401},
  issn    = {0003-004X},
  journal = {American Mineralogist},
  month   = {apr},
  number  = {3-4},
  pages   = {213--244},
  volume  = {82},
  year    = {1997}
}
@article{Koch:1984,
  author  = {Koch, E.},
  doi     = {10.1107/S0108767384001227},
  issn    = {0108-7673},
  journal = {Acta Crystallographica Section A Foundations of Crystallography},
  month   = {sep},
  number  = {5},
  pages   = {593--600},
  title   = {The implications of normalizers on group–subgroup relations between space groups},
  volume  = {40},
  year    = {1984}
}
```

and the .bibtoolrsc with a custom semantic check for the field months (biblatex wants you to use an integer value between 1 and 12):

```
% Semantic checks
check.rule { month "^[\"{]1?[0-9][\"}]$" }
check.rule { month "" "\@ \$: Month has to be a suitable number\n" }
```

I get follwoing message
```
❯ bibtool -r .bibtoolrsc my.bib > /dev/null
*** BibTool: Article koch:1984: Month has to be a suitable number
*** BibTool: Article dove:1997: Month has to be a suitable number
*** BibTool: Article koch:1984: Month has to be a suitable number
```

The results of bibclean for this file is
```
❯ bibclean my.bib > /dev/null
%% my.bib:3:Expected http://dx.doi.org/ prefix in DOI value ``"10.1107/S0108767384001227"''.
%% my.bib:16:Expected http://dx.doi.org/ prefix in DOI value ``"10.2138/am-1997-3-401"''.
%% my.bib:28:Expected http://dx.doi.org/ prefix in DOI value ``"10.1107/S0108767384001227"''.
```

which I find not very helpful.

Additionally, I would like to get a warning for duplicates without commenting or deletion, just  a message. The PR https://github.com/ge-ne/bibtool/pull/45 could improve the detection. But as I said I would like to get a message as well.

I checked already more tools than `bibclean` and `BibTool`:

- `biber -V` does not report line numbers as well. However, this seems to be a design choice and will most likely not change. The messages are very good out of the box and therefore is a pity that it cannot be used this way:

```
❯ biber --tool -V my.bib
INFO - This is Biber 2.11 running in TOOL mode
INFO - Logfile is 'my.bib.blg'
INFO - Looking for bibtex format file 'my.bib'
INFO - LaTeX decoding ...
INFO - Found BibTeX data source 'my.bib'
WARN - Duplicate entry key: 'Koch:1984' in file 'my.bib', skipping ...
WARN - month field 'apr' in entry 'Dove:1997' is not an integer - this will probably not sort properly.
WARN - month field 'sep' in entry 'Koch:1984' is not an integer - this will probably not sort properly.
WARN - BibTeX subsystem: /var/folders/d_/5kjxh9v92k752bvzf559fd3c0000gn/T/hvsk1NCjJd/my.bib_45262.utf8, line 13, warning: 1 characters of junk seen at toplevel
INFO - Overriding locale 'en_US' defaults 'variable = shifted' with 'variable = non-ignorable'
INFO - Overriding locale 'en_US' defaults 'normalization = NFD' with 'normalization = prenormalized'
INFO - Sorting list 'tool/global//global/global' of type 'entry' with template 'tool' and locale 'en_US'
INFO - No sort tailoring available for locale 'en_US'
INFO - Writing 'my_bibertool.bib' with encoding 'UTF-8'
INFO - Output to my_bibertool.bib
INFO - WARNINGS: 4
```
- [BibLatex-Check](https://github.com/Pezmc/BibLatex-Check) does not print line numbers as well and requires alwasy that the aux file exists.

**Comment by kiryph** [comment]
I want to add a list of possible checks which I might extend later on:

- duplicate records
- duplicate fields
- missing mandatory fields
- incorrect values (not integers for months, ...)

**Comment by ge-ne** [comment]
BibTool already uses messages of the following formats:
```
*** BibTool ERROR:  (line 3 in ./z.bib): Unexpected character encountered
*** BibTool WARNING:  (line 5 in ./z.bib): 17 non-space characters ignored.
```
But this works for parsing problems only since the line number is gone afterwards. Semantic checks are applied after the parsing...
BibTool already keeps the source file of a record. I could add the line number to the record. But I suspect that this is not possible to have reasonable values under any circumstance.
I would keep the line number of the starting line only -- for the whole record and maybe for each field. Since BibTeX can be formatted rather arbitrary this can be off several (dozen) lines. I already have the line number of the terminals but I don't keep them.

I will look into it. Right now it seems doable.

Concerning your checks.
Duplicate records are already recognized. Just the action needs to be added.
Duplicate fields are already treated. The last one wins;-) I will add an error message.
Mandatory fields are similar to record type definitions. The interesting part is that some fields are not just mandatory by themself.
Incorrect values can be covered with semantic checks. Note that month don't need to be numeric.

**Comment by kiryph** [comment]
> BibTool already uses messages of the following formats:

Surprisingly, I did not run into messages like this. But it is great to see this because they are suitable to be shown in the quickfix list in vim and others.

> BibTool already keeps the source file of a record. I could add the line number to the record. But I suspect that this is not possible to have reasonable values under any circumstance.
I would keep the line number of the starting line only -- for the whole record and maybe for each field. 

I think it still would be helpful to jump to the beginning of the record/field under consideration. No need to store more than this (IMHO).

> Duplicate records are already recognized. Just the action needs to be added.

Do you mean by action a message?

> Duplicate fields are already treated. The last one wins;-) I will add an error message.

This would be helpful. (ok it could be a warning not an error message 😄 )

> Mandatory fields are similar to record type definitions. The interesting part is that some fields are not just mandatory by themself.

Agreed.

> Incorrect values can be covered with semantic checks. Note that month don't need to be numeric.

True, only biblatex prefers numeric values and biber warns about that a chronological ordering might be incorrect.

**Comment by kiryph** [comment]
Thank you for your recent commits. I tried them:

1. Check Rules (018584c34246264775ab58369f5ace36155872a2):
```
❯ cat bibtoolrsc
% sort = on
% sort.format              = { $type year month {%N(author) # %N(editor)} }
% Semantic checks
check.rule { month "^[\"{]1?[0-9][\"}]$" }
check.rule { month "" "\@ \$: Month has to be a suitable number\n" }

❯ cat -n my.bib
     1	@article{Koch:1984,
     2	  author  = {Koch, E.},
     3	  doi     = {10.1107/S0108767384001227},
     4	  issn    = {0108-7673},
     5	  journal = {Acta Crystallographica Section A Foundations of Crystallography},
     6	  month   = {sep},
     7	  number  = {5},
     8	  pages   = {593--600},
     9	  title   = {{The implications of normalizers on group–subgroup relations between space groups}},
    10	  volume  = {40},
    11	  year    = {1984}
    12	}
    13	@Article{Dove:1997,
    14	  author  = {Dove, Martin T.},
    15	  title   = {{Theory of displacive phase transitions in minerals}},
    16	  doi     = {10.2138/am-1997-3-401},
    17	  issn    = {0003-004X},
    18	  journal = {American Mineralogist},
    19	  month   = {apr},
    20	  number  = {3-4},
    21	  pages   = {213--244},
    22	  url     = {https://pubs.geoscienceworld.org/ammin/article/82/3-4/213-244/43266},
    23	  volume  = {82},
    24	  year    = {1997}
    25	}
    26	@article{Koch:1984,
    27	  author  = {Koch, E.},
    28	  doi     = {10.1107/S0108767384001227},
    29	  issn    = {0108-7673},
    30	  journal = {Acta Crystallographica Section A Foundations of Crystallography},
    31	  month   = {sep},
    32	  number  = {5},
    33	  pages   = {593--600},
    34	  title   = {{The implications of normalizers on group–subgroup relations between space groups}},
    35	  volume  = {40},
    36	  year    = {1984}
    37	}

❯ /usr/local/Cellar/bib-tool/HEAD-d4abcda/bin/bibtool -r bibtoolrsc my.bib > /dev/null
*** BibTool (line 26 in my.bib): Article koch:1984: Month has to be a suitable number
*** BibTool (line 13 in my.bib): Article dove:1997: Month has to be a suitable number
*** BibTool (line 1 in my.bib): Article koch:1984: Month has to be a suitable number
```

Well, the messages are in reverse order, but that is not a problem. However, I am surprised about following behavior:

If I enable sorting, the messages disappear:
```
❯ cat bibtoolrsc
sort = on
sort.format              = { $type year month {%N(author) # %N(editor)} }
% Semantic checks
check.rule { month "^[\"{]1?[0-9][\"}]$" }
check.rule { month "" "\@ \$: Month has to be a suitable number\n" }

❯ /usr/local/Cellar/bib-tool/HEAD-d4abcda/bin/bibtool -r bibtoolrsc my.bib > /dev/null
```

Is this intended?

2. Duplicate record detection

I get as wished a message:
```
❯ cat my.bib
@article{Koch:1984,
  author  = {Koch, E.},
  doi     = {10.1107/S0108767384001227},
  issn    = {0108-7673},
  journal = {Acta Crystallographica Section A Foundations of Crystallography},
  month   = {sep},
  number  = {5},
  pages   = {593--600},
  title   = {{The implications of normalizers on group–subgroup relations between space groups}},
  volume  = {40},
  year    = {1984}
}
@Article{Dove:1997,
  author  = {Dove, Martin T.},
  title   = {{Theory of displacive phase transitions in minerals}},
  doi     = {10.2138/am-1997-3-401},
  issn    = {0003-004X},
  journal = {American Mineralogist},
  month   = {apr},
  number  = {3-4},
  pages   = {213--244},
  url     = {https://pubs.geoscienceworld.org/ammin/article/82/3-4/213-244/43266},
  volume  = {82},
  year    = {1997}
}
@article{Koch:1984,
  author  = {Koch, E.},
  doi     = {10.1107/S0108767384001227},
  issn    = {0108-7673},
  journal = {Acta Crystallographica Section A Foundations of Crystallography},
  month   = {sep},
  number  = {5},
  pages   = {593--600},
  title   = {{The implications of normalizers on group–subgroup relations between space groups}},
  volume  = {40},
  year    = {1984}
}

❯ /usr/local/Cellar/bib-tool/HEAD-d4abcda/bin/bibtool -d my.bib > /dev/null
*** BibTool WARNING: Possible double entries discovered:
***	koch:1984 =?= koch:1984
***	Article1984monthKoch.E

```

Again I would like to have a line number(s). Also the third line looks unintended.

3. Duplicate field detection (d4abcda4f9bb71fd8add5ac5172fc49e1a8d84ea)

I do not get a message in following circumstance. BibTool silently removes the duplicated month:

```
❯ cat my-dupfield.bib
@article{Koch:1984,
  author  = {Koch, E.},
  doi     = {10.1107/S0108767384001227},
  issn    = {0108-7673},
  journal = {Acta Crystallographica Section A Foundations of Crystallography},
  month   = {sep},
  number  = {5},
  pages   = {593--600},
  title   = {{The implications of normalizers on group–subgroup relations between space groups}},
  volume  = {40},
  month   = {sep},
  year    = {1984}
}

❯ /usr/local/Cellar/bib-tool/HEAD-d4abcda/bin/bibtool my-dupfield.bib
@Article{Koch:1984,
  author        = {Koch, E.},
  doi           = {10.1107/S0108767384001227},
  issn          = {0108-7673},
  journal       = {Acta Crystallographica Section A Foundations of
                  Crystallography},
  month         = {sep},
  number        = {5},
  pages         = {593--600},
  title         = {{The implications of normalizers on group–subgroup
                  relations between space groups}},
  volume        = {40},
  year          = {1984}
}
```
Adding `-d` or `-v` do not show me the message. What do I have to do?

**Comment by ge-ne** [comment]
> However, I am surprised about following behavior:
> If I enable sorting, the messages disappear:
> Is this intended?

I was not able to reproduce this behaviour. I would be glad to receive more details.

**Comment by kiryph** [comment]
> I would be glad to receive more details.

I already posted the `bibtoolrsc` and the bib file `my.bib` and I tried the most recent commit.
What else do you need? 

OS: macOS 10.14.2
File system: apfs (case insensitive)
Shell: zsh 5.6.2 (x86_64-apple-darwin16.7.0)

**Comment by kiryph** [comment]
I ran the test suite as well. Everything is ok except following which seems to be unrelated:
```
  keep_field_01...........................
***	See actual out in keep_field_01.out;
				  fail
  keep_field_02...........................
***	See actual out in keep_field_02.out;
				  fail
  keep_field_11...........................
***	See actual out in keep_field_11.out;
				  fail
  keep_field_21...........................
***	See actual out in keep_field_21.out;
				  fail
  keep_field_22...........................
***	See actual out in keep_field_22.out;
				  fail
  keep_field_23...........................
***	See actual out in keep_field_23.out;
				  fail
```
```
❯ diff keep_field_01.out-expected keep_field_01.out
2,3c2
< @Article{	  bibtool,
<   title		= {The BibTool Manual}
---
> @Article{	  bibtool
6,7c5
< @Manual{	  bibtool2,
<   title		= {The BibTool Manual}
---
> @Manual{	  bibtool2
```

Warnings
```
Use of uninitialized value $ENV{"HOSTNAME"} in concatenation (.) or string at ./key_format.t line 2602.
*** Warning: Multiple test cases named check_double_1
```

**Comment by kiryph** [comment]
New compiler warnings between commit c71d56c and b93b53aa93ba450678194a2424e5034343cf18c0
```diff
diff --git a/warnings-old b/warnings-new
index 456ef8b..8b1b8c3 100644
--- a/warnings-old
+++ b/warnings-new
@@ -6,7 +6,24 @@ main.c:173:18: warning: promoted type 'int' of K&R function parameter is not com
 main.c:94:31: note: previous declaration is here
  static void usage _ARG((bool fullp));             /* main.c                 */
                               ^
-1 warning generated.
+main.c:797:6: warning: passing 'char [36]' to parameter of type 'String' (aka 'unsigned char *') converts between pointers to integer types
+      with different sign [-Wpointer-sign]
+            "Possible double entry discovered to", /*                        */
+            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+include/bibtool/error.h:458:35: note: passing argument to parameter 's1' here
+ void error _ARG((int type,String s1,String s2,String s3,String line,String ep,int line_no,char *file_name));/* error.c*/
+                                  ^
+main.c:799:25: warning: passing 'Symbol' (aka 'unsigned char *') to parameter of type 'char *' converts between pointers to integer types with
+      different sign [-Wpointer-sign]
+            RecordLineno(rec), RecordSource(rec)); /*                        */
+                               ^~~~~~~~~~~~~~~~~
+include/bibtool/record.h:359:25: note: expanded from macro 'RecordSource'
+#define RecordSource(R) ((R)->rc_source)
+                        ^~~~~~~~~~~~~~~~
+include/bibtool/error.h:458:98: note: passing argument to parameter 'file_name' here
+ void error _ARG((int type,String s1,String s2,String s3,String line,String ep,int line_no,char *file_name));/* error.c*/
+                                                                                                 ^
+3 warnings generated.
 gcc -g -O2   -Iinclude  -c -c crossref.c -o crossref.o
 crossref.c:352:37: warning: passing 'char [9]' to parameter of type 'String' (aka 'unsigned char *') converts between pointers to integer
       types with different sign [-Wpointer-sign]
@@ -129,7 +146,50 @@ gcc -g -O2   -Iinclude   -c parse.c -o parse.o
 gcc -g -O2   -Iinclude  -c -c print.c -o print.o
 gcc -g -O2   -Iinclude  -c -c pxfile.c -o pxfile.o
 gcc -g -O2   -Iinclude  -c -c record.c -o record.o
+record.c:334:8: warning: promoted type 'int' of K&R function parameter is not compatible with the parameter type 'bool' declared in a previous
+      prototype [-Wknr-promoted-parameter]
+  bool err;                                        /*                        */
+       ^
+record.c:52:63: note: previous declaration is here
+ void push_to_record _ARG((Record rec,Symbol s,Symbol t, bool err));/* record.c*/
+                                                              ^
+record.c:346:8: warning: passing 'char [18]' to parameter of type 'String' (aka 'unsigned char *') converts between pointers to integer types
+      with different sign [-Wpointer-sign]
+              "Duplicate field `", s, "' overwritten",/*                     */
+              ^~~~~~~~~~~~~~~~~~~
+include/bibtool/error.h:458:35: note: passing argument to parameter 's1' here
+ void error _ARG((int type,String s1,String s2,String s3,String line,String ep,int line_no,char *file_name));/* error.c*/
+                                  ^
+record.c:346:32: warning: passing 'char [14]' to parameter of type 'String' (aka 'unsigned char *') converts between pointers to integer types
+      with different sign [-Wpointer-sign]
+              "Duplicate field `", s, "' overwritten",/*                     */
+                                      ^~~~~~~~~~~~~~~
+include/bibtool/error.h:458:55: note: passing argument to parameter 's3' here
+ void error _ARG((int type,String s1,String s2,String s3,String line,String ep,int line_no,char *file_name));/* error.c*/
+                                                      ^
+record.c:348:27: warning: passing 'Symbol' (aka 'unsigned char *') to parameter of type 'char *' converts between pointers to integer types
+      with different sign [-Wpointer-sign]
+              RecordLineno(rec), RecordSource(rec)); /*                      */
+                                 ^~~~~~~~~~~~~~~~~
+include/bibtool/record.h:359:25: note: expanded from macro 'RecordSource'
+#define RecordSource(R) ((R)->rc_source)
+                        ^~~~~~~~~~~~~~~~
+include/bibtool/error.h:458:98: note: passing argument to parameter 'file_name' here
+ void error _ARG((int type,String s1,String s2,String s3,String line,String ep,int line_no,char *file_name));/* error.c*/
+                                                                                                 ^
+4 warnings generated.
 gcc -g -O2   -Iinclude  -DREGEX -Iregex-0.12 -I.. -c rewrite.c -o rewrite.o
+rewrite.c:949:27: warning: passing 'Symbol' (aka 'unsigned char *') to parameter of type 'char *' converts between pointers to integer types
+      with different sign [-Wpointer-sign]
+              RecordLineno(rec), RecordSource(rec));/*                       */
+                                 ^~~~~~~~~~~~~~~~~
+include/bibtool/record.h:359:25: note: expanded from macro 'RecordSource'
+#define RecordSource(R) ((R)->rc_source)
+                        ^~~~~~~~~~~~~~~~
+include/bibtool/error.h:458:98: note: passing argument to parameter 'file_name' here
+ void error _ARG((int type,String s1,String s2,String s3,String line,String ep,int line_no,char *file_name));/* error.c*/
+                                                                                                 ^
+1 warning generated.
 gcc -g -O2   -Iinclude  -DRSC_BIBINPUTS_DEFAULT=NULL -DRSC_BIBTOOL_DEFAULT=\".:/usr/local/Cellar/bib-tool/HEAD-b93b53a/lib/BibTool\" -DENV_SEP=\":\" -c rsc.c -o rsc.o
 gcc -g -O2   -Iinclude  -c -c s_parse.c -o s_parse.o
 s_parse.c:190:11: warning: promoted type 'int' of K&R function parameter is not compatible with the parameter type 'bool' declared in a
```

**Comment by ge-ne** [comment]
> > I would be glad to receive more details.

> I already posted the bibtoolrsc and the bib file my.bib and I tried the most recent commit.
> What else do you need?

I have a test case to show this effect too. I will work on it.

**Comment by kiryph** [comment]
Yes, I can confirm that `check_rule_6` in `test/check_rule.t` does not succeed. The output is empty:
```
❯ cat check_rule_6.err

❯ cat check_rule_6.err-expected
*** BibTool (line 3 in <STDIN>): Manual bibtool: Year has to be a suitable number
```

**Comment by kiryph** [comment]
1. **Check rule messages are back when used together with sorting**
The messages are not missing anymore (presumably due to 2938cc88c04791dd4c0fdb838296fd9653cf6b5b):
```
❯ cat bibtoolrsc
sort = on
sort.format              = { $type year month {%N(author) # %N(editor)} }
% Semantic checks
check.rule { month "^[\"{]1?[0-9][\"}]$" }
check.rule { month "" "\@ \$: Month has to be a suitable number\n" }

~
❯ /usr/local/Cellar/bib-tool/HEAD-96e2dde/bin/bibtool -r bibtoolrsc my.bib > /dev/null
*** BibTool (line 26 in my.bib): Article koch:1984: Month has to be a suitable number

*** BibTool (line 13 in my.bib): Article dove:1997: Month has to be a suitable number

*** BibTool (line 1 in my.bib): Article koch:1984: Month has to be a suitable number

```
However, I am wondering whether the additional blank lines are necessary?

2. **New default behavior w.r.t. stdout of bib file**
I also noticed that you changed the default behavior of bibtool in commit edaa99baa68df575ad7056bdcdc86f03dc96b0ca whether the bib file is printed to stdout. I am not sure whether existing user appreciate this change. Adding a single minus sign is not a big deal but still it is a breaking change.

3. **Incomplete duplicates message**
I also encountered an issue with the message for duplicates

```
❯ /usr/local/Cellar/bib-tool/2.67/bin/bibtool -s -d  my.bib > /dev/null
*** BibTool WARNING: Possible double entries discovered:
***	koch:1984 =?= koch:1984
***	koch:1984

❯ /usr/local/Cellar/bib-tool/HEAD-96e2dde/bin/bibtool -s -d  my.bib > /dev/null
 (line 1 in my.bib) `koch:1984'
```
The message is incomplete.

**Comment by ge-ne** [comment]
> 1. Check rule messages are back when used together with sorting

> However, I am wondering whether the additional blank lines are necessary?

They are useful when the progress report is turned on. With some effort they can be eliminated. But this does not have high priority for me right now.

> 1. New default behavior w.r.t. stdout of bib file

The novelty is that an empty string can be used to suppress any output. The behaviour for - has been in place for decades. I have never had any questions or complaints about it. Right now I have prominently added documentation for this features. Thus I do not consider this as a breaking change.

> 2. Incomplete duplicates message
[...]
> The message is incomplete.

Fixed

**Comment by kiryph** [comment]
> But this does not have high priority for me right now.

This is indeed a minor thing.

> The novelty is that an empty string can be used to suppress any output. The behaviour for - has been in place for decades. I have never had any questions or complaints about it.

Actually, I got it wrong. I was thinking
```
$ bibtool -s ~/my.bib
```
would now return nothing. Only
```
$ bibtool -s -o "" ~/my.bib
```
returns now only messages. A little bit shorter than redirecting to `/dev/null` and actually quite intuitve even though I got it wrong in the first place.


> Incomplete duplicates message
[...]
The message is incomplete.
Fixed

Thank you.

I think as a next step I will create a compiler plugin for vim. I will check whether I get all the messages I was hoping to see within vim when having a bib file with duplicates and a few issues which bibtool can spot and biber usually warns about.
